docs/reference/modules/scripting/fields.asciidoc

```diff
@@ -0,0 +1,232 @@
+
+===================================================
+
+[float]
+[[modules-scripting-stored]]
+=== Stored Fields and `_source`
+
+_Stored fields_ -- fields explicitly marked as
+<<mapping-store,`"store": true`>> -- can be accessed using the
+`_fields['field_name'].value` or `_fields['field_name'].values` syntax.
+
+The document <<mapping-source-field,`_source`>>, which is really just a
+special stored field,  can be accessed using the `_source.field_name` syntax.
+The `_source` is loaded as a map-of-maps, so properties within object fields
+can be accessed as, for example, `_source.name.first`.
+
+[IMPORTANT]
+.Prefer doc-values to stored fields
+=========================================================
+
+Stored fields (which includes the stored `_source` field) are much slower than
+doc-values.  They are  optimised for returning several fields per result,
+while doc values are optimised for accessing the value of a specific field in
+many documents.
+
+
+It makes sense to use `_source` or stored fields when generating a
+<<search-request-script-fields,script field>> for the top ten hits from a search
+result but, for other search and aggregation use cases, always prefer using
+doc values.
+=========================================================
+
+
+For instance:
+
+[source,js]
+-------------------------------
+PUT my_index
+{
+  "mappings": {
+    "my_type": {
+      "properties": {
+        "title": { <1>
+          "type": "text"
+        },
+        "first_name": {
+          "type": "text",
+          "store": true
+        },
+        "last_name": {
+          "type": "text",
+          "store": true
+        }
+      }
+    }
+  }
+}
+
+PUT my_index/my_type/1
+{
+  "title": "Mr",
+  "first_name": "Barry",
+  "last_name": "White"
+}
+
+GET my_index/_search
+{
+  "script_fields": {
+    "source": {
+      "script": {
+        "lang": "groovy",
+        "inline": "_source.title + ' ' + _source.first_name + ' ' + _source.last_name" <2>
+      }
+    },
+    "stored_fields": {
+      "script": {
+        "lang": "groovy",
+        "inline": "_fields['first_name'].value + ' ' + _fields['last_name'].value"
+      }
+    }
+  }
+}
+-------------------------------
+// AUTOSENSE
+<1> The `title` field is not stored and so cannot be used with the `_fields[]` syntax.
+<2> The `title` field can still be accessed from the `_source`.
+
+[TIP]
+.Stored vs `_source`
+=======================================================
+
+The `_source` field is just a special stored field, so the performance is
+similar to that of other stored fields.  The `_source` provides access to the
+original document body that was indexed (including the ability to distinguish
+`null` values from empty fields, single-value arrays from plain scalars, etc).
+
+The only time it really makes sense to use stored fields instead of the
+`_source` field is when the `_source` is very large and it is less costly to
+access a few small stored fields instead of the entire `_source`.
```

Comment:
For the record another difference is that stored fields store a more primitive representation of the data (milliseconds since 1970 for dates for instance). I am not sure whether this is worth documenting.